sky/data/storage_utils.py

```diff
@@ -42,3 +46,98 @@ def format_storage_table(storages: List[Dict[str, Any]]) -> str:
+    """ Lists files and patterns ignored by git in the source directory
+
+    Runs `git status --ignored` which returns a list of excluded files and
+    patterns read from .gitignore and .git/info/exclude using git.
+    `git init` is run if SRC_DIR_PATH is not a git repository and removed
+    after obtaining excluded list.
+
+    Returns:
+        List[str] containing files and patterns to be ignored.  Some of the
+        patterns include, **/mydir/*.txt, !myfile.log, or file-*/.
+    """
+    expand_src_dir_path = os.path.expanduser(src_dir_path)
+
+    git_exclude_path = os.path.join(expand_src_dir_path, '.git/info/exclude')
+    gitignore_path = os.path.join(expand_src_dir_path, '.gitignore')
+
+    git_exclude_exists = os.path.isfile(git_exclude_path)
+    gitignore_exists = os.path.isfile(gitignore_path)
+
+    # This command outputs a list to be excluded according to .gitignore
+    # and .git/info/exclude
+    filter_cmd = f'git -C {expand_src_dir_path} status --ignored --porcelain=v1'
+    excluded_list: List[str] = []
+
+    if git_exclude_exists or gitignore_exists:
+        try:
+            output = subprocess.run(filter_cmd,
+                                    shell=True,
+                                    stdout=subprocess.PIPE,
+                                    stderr=subprocess.PIPE,
+                                    check=True,
+                                    text=True)
+        except subprocess.CalledProcessError as e:
+            # when the SRC_DIR_PATH is not a git repo and .git
+            # does not exist in it
+            if e.returncode == exceptions.GIT_FATAL_EXIT_CODE:
+                if 'not a git repository' in e.stderr:
+                    # Check if the user has 'write' permission to
+                    # SRC_DIR_PATH
+                    if not os.access(expand_src_dir_path, os.W_OK):
+                        logger.warning(
+                            f'{colorama.Fore.YELLOW}Warning: Write permission '
+                            f'denied in {src_dir_path}. Files/dirs specified'
+                            ' in .gitignore will be uploaded to the Cloud '
+                            'Storage."')
+                        return excluded_list
+                    init_cmd = f'git -C {expand_src_dir_path} init'
+                    subprocess.run(init_cmd,
+                                   shell=True,
+                                   stdout=subprocess.PIPE,
+                                   stderr=subprocess.PIPE,
+                                   check=True)
+                    output = subprocess.run(filter_cmd,
+                                            shell=True,
+                                            stdout=subprocess.PIPE,
+                                            stderr=subprocess.PIPE,
+                                            check=True,
+                                            text=True)
```

Comment:
These two commands _may_ fail unexpectedly. We should wrap them in a try catch block and handle any unexpected failures by falling back to ignoring contents of gitignore.